REVIEW: Refactor Git implementation, extract platform specific code

adrienthebo opened the pull request:
This pull request refactors the code interacting with Git. It pulls out the code that directly interacts with the Git binary (by shelling out, running commands, parsing the output, etc) into a separate set of classes.

These classes directly interact with Git:
- BareRepository
- WorkingRepository
- ThinRepository (a WorkingRepository that uses Git alternates to perform shallow clones)

These classes wrap interactions with Git but aren't bound to the Git implementation:
- Cache (wraps a BareRepository and handles updating that repository when necessary)
- StatefulRepository (wraps a ThinRepository and acts to synchronize a Git repository with some desired state for that repository)

The goal of this refactor is to break up the low level logic that interacts with the Git binary into a distinct layer that can be swapped out, and adds a logic layer that actually provides the application logic to run everything.

adrienthebo committed:
[spec] add git repo fixtures for Git integration

This commit adds fixture data and rspec helpers for performing
integration tests touching Git repositories. Trying to stub out every
single git command touching a git repository would be more work than
just running the commands and checking the result, so this should
simplify testing. Moreover this allows different Git implementations to
share the same tests to confirm they have the same behavior.

adrienthebo committed:
[git/bare_repository] (RK-16) Implement Git bare repository

This commit adds a generic implementation of a bare Git repository. The
existing code that deals with bare repositories is tightly coupled with
how to cache Git repositories and is hard to test it, so this class was
extracted to just deal with how Git bare repositories behave.

adrienthebo committed:
[git/cache] (RK-16) Use BareRepository from Git cache

The previous implementation of the Git caching code intermixed caching
logic and Git commands which made things hard to test and reason about.
This commit removes all of the Git implementation logic from the cache
class and reuses the BareRepository class, so the Cache only has to deal
with, well, caching.

adrienthebo committed:
[git/working_repository] (RK-16) Implement Git non-bare repo

The WorkingDir class implemented interactions with non-bare git
repositories, but it blended handling a non-bare repository, performing
sparse clones, and enforcing the state of a given repository. This
commit implements a class for just interacting with a non-bare
repository; the concerns of sparse clones and state enforcement will be
handled by separate classes.

adrienthebo committed:
[git/alternates] (RK-16) Use Pathname instead of string path

Pathname objects are significantly more convenient to use for dealing with
files and file paths and I've been replacing String path members with
Pathname members. This commit updates the Alternates object to use a
Pathname instead of a string object as well.

adrienthebo committed:
[git/thin_repository] (RK-16) Implement cache backed working repo

This class extends a working repository to add sparse clones and
updates. All clone and fetch operations are done by a cached git
repository and the actual working repository simply reuses the cache
repo object database.

adrienthebo committed:
[git/bare_repository] (RK 16) Add ref resolution

This adds the #resolve method to bare repositories to resolve git refs
to actual commits.

adrienthebo committed:
[git/bare_repository] (RK-16) Determine what type a ref is

This adds the #ref_type method that takes a gif ref and will return the
type of the ref.

adrienthebo committed:
[git/bare_repository] (RK-16) Remove remote from initialization field

The remote of a bare repository is the property of an existing repo and
is not an intrinsic part of the object initialization, so it cannot be
supplied to the constructor.

adrienthebo committed:
[git] (RK-16) convert #git_dir to always return Pathname

This was another location where file paths were being treated as Strings
instead of Pathnames; converted this to a Pathname as well for
consistency.

adrienthebo committed:
[git/working_repository] (RK-16) add #git_dir

This adds the #git_dir method to the WorkingRepository object, so that
ThinRepository, WorkingRepository, and BareRepository all implement the
same interface.

adrienthebo committed:
[git/base_repository] (RK-16) extract #git method to base class

The BareRepository and WorkingRepository classes both need to run git
commands but the initial versions of these classes just copied the old
`#git` method. This commit extracts a base class and adds the #git
method to that class.

adrienthebo committed:
[git/base_repository] (RK-16) extract #tags

This commit removes the duplicate #tags implementation from the
BareRepository and WorkingRepository classes and moves it to the base
class.

adrienthebo committed:
[git/base_repository] (RK-16) extract #resolve method

This commit removes the duplicate #resolve implementation from the
BareRepository and WorkingRepository classes and moves it to the base
class.

adrienthebo committed:
[git/base_repository] (RK-16) add abstract #git_dir definition

This implements an abstract `#git_dir` method in the BaseRepository to
indicate that subclasses should implement it.

adrienthebo committed:
[git/base_repository] (RK-16) centralize #branches method

This commit removes the duplicate #branches implementation from the
BareRepository and WorkingRepository classes and moves it to the base
class.

adrienthebo committed:
[working_repository] (RK-16) don't raise error on nil origin

The #origin method may need to be invoked before the repository is
initialized, so if the git command to retrieve the origin remote url
fails then just return nil.

adrienthebo committed:
[git/base_repository] (RK-16) extract #for_each_ref helper

Enumerating tags and branches requires invoking `git for-each-ref` with
slightly different args; this centralizes the shared behavior.

adrienthebo committed:
[working_repository] (RK-16) Use #checkout method after initial clone

WorkingRepository instances could clone a repository and check out an
initial ref, but `git clone -b` can only take a branch or tag and not a
commit. In order to resolve this consistency and make
branches/tags/commit be equivalent this commit uses #checkout which
treats the three as equivalent.

adrienthebo committed:
[git/cache] (RK-16) delegate #resolve to bare repo

The Cache object may be used as if it is a git repository itself; to
facilitate that we just forward the #resolve method to the backing bare
repository.

adrienthebo committed:
[git/stateful_repository] (RK-16) Add repo manager class

This class controls how Git repositories are created and updated. Class
instances take a desired Git remote and reference, compare those states
to the repository on disk, and can reconcile the two. It takes advantage
of cached Git repositories to deduplicate Git objects and speed up
operations.

adrienthebo committed:
[git] (RK-16) Swap environment/module Git implementation

The R10K::Environment::Git and R10K::Module::Git classes shared a lot of
behavior on how repositories are synchronized that have been subsumed
into the StatefulRepository class. We can replace the duplicate code
with calls into the stateful repo instead.

adrienthebo committed:
[git/cache] (RK-16) Forward ref_type call to bare repo

adrienthebo committed:
[git/cache] (RK-16) Expose sync status, allow status reset

adrienthebo committed:
[git/stateful_repository] (RK-16) Treat unsynced branches as outdated

Git branches are a bit strange in comparison to Git commit SHAs and tags
because they are mutable. If we've been given a Git ref that's a branch,
we have to make sure that the branch itself has been fetched so that we
actually track upstream changes to the branch.

adrienthebo committed:
[git/stateful_repository] (RK-16) Raise error on unresolvable ref

If the desired ref cannot be resolved to a SHA, we cannot synchronize
the repository and must fail immediately.

adrienthebo committed:
[git/stateful_repository] (RK-16) forward #head to backing repo

adrienthebo committed:
[module/git] (RK-16) Remove use of ref objects

The R10K::Git::Ref classes and subclasses were an attempt to add some
optimizations to how r10k fetches and updates Git repos, but their
architecture was flawed and people weren't widely using them. The new
Git classes don't use these classes so we can just remove the classes
and use the string refs instead.

adrienthebo committed:
[git/base_repository] (RK-16) Move #ref_type to base repo

This method is common to all Git repositories; it can be shifted to the
Git base class.

adrienthebo committed:
[spec] (maint) Expand remote filepath for cross platform compat

OSX sets the TMPDIR path to `/var/folders/[randomstring]`, but `/var` is
a symlink to `/private/var`, and it looks like Git will dereference
paths given via `git clone --reference [filepath]`. To make sure the
specs pass on any platform we expand the fixture path as well.

andersonmills commented on lib/r10k/module/git.rb:
```diff
@@ -29,69 +30,41 @@ def version
-    return :insync
+    @repo.status
   end
 
```
Any reason you didn't use Forwardable and def_delegators for sync and status as in cache.rb above?

andersonmills commented on spec/integration/git/working_repository_spec.rb:
```diff
@@ -0,0 +1,97 @@
+    end
+
+    it "lists the local branches in alphabetical order" do
+      expect(subject.branches).to eq(%w[master])
```
I'd say that the name of this test is a little disingenuous for what the test tests.

adrienthebo committed:
[git] (RK-16) Use def_delegators to forward method calls

Before this commit, the Git environment and module had moved most of
their behavior to the backing repo class, but did that delegation by
calling the delegated method. A number of other places were using
`def_delegators` to provide this behavior instead. While it's a small
thing, the environment and module classes were inconsistent.

This commit updates the Git environment and module classes to use
`def_delegator` for methods that are purely delegated to the backing
repo class.
